test/cli/commands_test.py

```diff
@@ -509,6 +511,162 @@ def test__cli__command__fix(rule, fname):
+            """,
+            ["--force", "--fixed-suffix", "FIXED", "--rules", "L010"],
+            None,
+            1,
+        ),
+        (
+            # - One lint error: "where" is lower case
+            # - Not fixable because of parse error (even though "noqa"), hence
+            #   error exit
+            """
+            SELECT my_col
+            FROM my_schema.my_table
+            where processdate ! 3  -- noqa: PRS
+            """,
+            ["--force", "--fixed-suffix", "FIXED", "--rules", "L010"],
+            None,
+            1,
+        ),
+        (
+            # - No lint errors
+            # - Parse error not suppressed, hence error exit
+            """
+            SELECT my_col
+            FROM my_schema.my_table
+            WHERE processdate ! 3
+            """,
+            ["--force", "--fixed-suffix", "FIXED", "--rules", "L010"],
+            None,
+            1,
+        ),
+        (
+            # - No lint errors
+            # - Parse error suppressed, hence success exit
+            """
+            SELECT my_col
+            FROM my_schema.my_table
+            WHERE processdate ! 3  --noqa: PRS
+            """,
+            ["--force", "--fixed-suffix", "FIXED", "--rules", "L010"],
+            None,
+            0,
+        ),
+        (
+            # - One lint error: "where" is lower case
+            # - Parse error not suppressed
+            # - "--FIX-EVEN-UNPARSABLE", hence fix anyway & success exit
+            """
+            SELECT my_col
+            FROM my_schema.my_table
+            where processdate ! 3
+            """,
+            [
+                "--force",
+                "--fixed-suffix",
+                "FIXED",
+                "--rules",
+                "L010",
+                "--FIX-EVEN-UNPARSABLE",
+            ],
+            """
+            SELECT my_col
+            FROM my_schema.my_table
+            WHERE processdate ! 3
+            """,
+            0,
+        ),
+        (
+            # Two files:
+            # File #1:
+            #   - One lint error: "where" is lower case
+            #   - Not fixable because of parse error
+            # File #2:
+            #   - One lint error: "where" is lower case
+            #   - No parse error, thus fixable
+            # Should fix the second file but not the first, and exit with an
+            # error.
+            [
+                """
+                SELECT my_col
+                FROM my_schema.my_table
+                where processdate ! 3
+                """,
+                """SELECT my_col
+                FROM my_schema.my_table
+                where processdate != 3""",
+            ],
+            ["--force", "--fixed-suffix", "FIXED", "--rules", "L010"],
+            [
+                None,
+                """SELECT my_col
+                FROM my_schema.my_table
+                WHERE processdate != 3""",
+            ],
+            1,
+        ),
+    ],
```

Comment:
These test cases are worth a careful review. They should match with the behavior we discussed in the maintainers meeting and added to issue #2424:

`sqlfluff fix`:
- Only fix if no unparsable segments found in the file. If any are found, no fixes happen even if parsing errors are supressed using noqa or `--ignore` cli options. The only way to override this is with the `--FIX-EVEN-UNPARSABLE` option as above.
- If there are still parse errors after applying "noqa" directives, exit 1. Otherwise, allow exit 0.

Reply:
I think these test cases reflect what we intended 👍 .